gateways/metadata_gateway.py

```diff
@@ -0,0 +1,80 @@
+import json
+from typing import Optional
+
+from common.configuration import METADATA_BUCKET
+from domain.metadata.metadata import MetadataType
+from domain.metadata.token_metadata import TokenMetadata
+from domain.metadata.transaction_metadata import TransactionMetadata
+from gateways.clients.s3_client import S3Client
+
+TYPE_TO_FOLDER = {
+    MetadataType.TRANSACTION: 'transaction',
+    MetadataType.TOKEN: 'token'
+}
+
+
+class MetadataGateway:
+
+    def __init__(self, s3_client: Optional[S3Client] = None) -> None:
+        self.s3_client = s3_client or S3Client()
+
+    def get_transaction_metadata(self, id: str) -> Optional[TransactionMetadata]:
+        """Retrieve transaction metadata from a json file stored in s3
+
+        :param id: transaction id
+        :type id: str
+        :raises Exception: The name of the bucket used to store the jsons must be on config
+        :return: TransactionMetadata object or None if not found
+        :rtype: TokenMetadata | None
+        """
+        folder = TYPE_TO_FOLDER[MetadataType.TRANSACTION]
+        transaction_metadata = self._get_metadata(f"{folder}/{id}.json")
+
+        if transaction_metadata is None:
+            return None
+
+        return TransactionMetadata.from_dict(transaction_metadata)
+
+    def get_token_metadata(self, id: str) -> Optional[TokenMetadata]:
```

Comment:
This method and the one above are exactly the same, right? Maybe we can create a generic method that receives `id`, `folder` and `cls` (the class to call `from_dict`).

Reply:
`_get_metadata` is already generic. It receives an s3 object path. Those two methods just calculate this paths and returns it as the right class. They are very similar because their jsons are stored in the same way. The ideia is each one know how to calculate your own object path, use `_get_metadata` (or maybe another method) to retrieve it and return the right class having the data